added Hw 2

maxkvant opened the pull request.

maxkvant committed:
added feature extraction

maxkvant committed:
incorporate feature extraction into sampler

maxkvant committed:
added rnn

maxkvant committed:
added rnn training and plots

maxkvant committed:
added plots for proba

maxkvant committed:
added pr plot; spedup feature extraction

maxkvant committed:
added model saving

maxkvant committed:
model saves

maxkvant committed:
fixed memory leak, changed noised dataset

maxkvant committed:
added models

maxkvant committed:
fix

maxkvant committed:
fix

Rauf-Kurbanov commented on 2/homework/Makefile:
```diff
@@ -0,0 +1,7 @@
+test:
+    py.test tests
+
+.PHONY: init test
```
It is easier to maintain putting .PHONY right before the target

Rauf-Kurbanov reviewed (changes requested):
Nice job, a lot of plots.

Common notice: 
1) no eof newlines
2) low f1 score alongside with high auc suggests poorly selected threshold or imbalanced validation set. 

To make this possible to review:
1) please eliminate all unused code from PR
2) Split your PR in two: in first merge all copy-pasted files [no need to review], in second: add your work on top of this so I can review it

maxkvant committed:
added spaces at ends of the files

maxkvant commented:
> 1. no eof newlines

a new commit for it

> 2. Split your PR in two: in first merge all copy-pasted files [no need to review], in second: add your work on top of this so I can review it

just merged copy-pasted files into master

Rauf-Kurbanov commented:
@maxkvant as a general comment please re-request review not only in email but also on a github

Rauf-Kurbanov commented on 2/homework/laughter_classification/sspnet_data_sampler.py:
```diff
@@ -88,14 +108,27 @@ def create_sampled_df(self, frame_sec, naudio=None, save_path=None, force_save=F
+        start_time = time.time()
+
+        def read_dataframe(self_iter_path_frame_n):
+            self_, iter, wav_path, frame_sec, n = self_iter_path_frame_n
```
`iter` is bad variable name

Rauf-Kurbanov commented on 2/homework/laughter_classification/sspnet_data_sampler.py:
```diff
@@ -88,14 +108,27 @@ def create_sampled_df(self, frame_sec, naudio=None, save_path=None, force_save=F
-        colnames.append("IS_LAUGHTER")
-        colnames.append("SNAME")
-        df.columns = colnames
+        start_time = time.time()
```
You might want to move start_time into `read_dataframe ` funcion fro more consistency
Another way to do this might be using a context manager
https://preshing.com/20110924/timing-your-code-using-pythons-with-statement/

maxkvant replied:
`read_dataframe` is a helper for parallel feature computing.
I don't want to move it, because start time means the time before start of parallel computing.

Rauf-Kurbanov commented on 2/homework/laughter_classification/sspnet_data_sampler.py:
```diff
@@ -88,14 +108,27 @@ def create_sampled_df(self, frame_sec, naudio=None, save_path=None, force_save=F
+            res = self_.df_fbank_mfcc_from_file(wav_path, frame_sec)
+
+            if (iter > 2) and (iter & (iter - 1) == 0):
+                print("iter {}/{} {}".format(iter, n, time.time() - start_time))
```
Explore f-strings
https://www.python.org/dev/peps/pep-0498/

Rauf-Kurbanov commented on 2/homework/laughter_prediction/rnn.py:
```diff
@@ -0,0 +1,39 @@
+        self.linear_mfcc = nn.Linear(hidden_size, 2)
+
+    def init_hidden_state(self, device):
+        return (torch.randn(1, self.batch_size, self.hidden_size).to(device),
```
In pytorch you don't need to specify device for particular layers since you can move your whole nn.Module to device all at once

Rauf-Kurbanov commented on 2/homework/laughter_prediction/predictors.py:
```diff
@@ -74,9 +77,18 @@ def __init__(self, threshold=0.045985743):
+        self.model.set_batch_size(len(X))
+        X_tensor = torch.tensor(X).float().to(device)
+        predicted_proba, _ = self.model.forward(X_tensor, device)
+        return torch.exp(predicted_proba[:,:,1]).cpu().detach().numpy()
```
I realise this might suggested by my poor choice of predictor interface. Just keep in mind that in pytorch it is ok to return logits logarithms without exponentiating. Also preserves numeric stability

Rauf-Kurbanov commented on 2/homework/laughter_prediction/rnn.py:
```diff
@@ -0,0 +1,39 @@
+        x_fbank = x[:, :, :self.fbank_features].permute(1, 0, 2)
+        x_mfcc  = x[:, :, self.fbank_features:].permute(1, 0, 2)
+
+        out_fbank_lstm, _ = self.fbank_lstm(x_fbank, self.init_hidden_state(device))
```
@maxkvant Why do you always initialize lstm hidden state with random value when you should pass the hidden from previous pass?

maxkvant replied:
I know 4 options for lstm initializartion:
1. zero state
2. random state
3. state from previous pass
4. noised state from previous pass

I have doubts about how to do approach 3 correctly for batches (Should I average it or not?). That's why I didn't consider options 3 and 4.

According to [post](https://r2rt.com/non-zero-initial-states-for-recurrent-neural-networks.html) random state better than zero state.

Rauf-Kurbanov reviewed (changes requested):
General comments:
* Some minor violations of pep8
* Do not upload your checkpoints in repo, since it will become huge. Also take into account that you will have to carry this files in history even after removal. As an alternative you can wget from dropbox or use git-lfs.
* Left some minor suggestions on a code

On the notebook:
* You load `autoreload` plugin but don't use it
* You can add `;` after plot function to avoid unnecessary output
* Please do normalise you features !
* Adam is more safe optimisator choice if you are not planning to spend a lot of time on fine-tuning and lr-scheduling
* For general metrics you might want to use `sklearn.metric`
* Why would you need `global` in jupyter?
* Tensorboard for metrics is much better practice than pyplot. It also take care of smoothing
* Too much code in jupyter, it is ok for a homework but it is hard to review and maintain in real life.

Summary: please answer my question about hidden state and fix feature normalisation.
Other comments are just for you, can ignore them if you wish

maxkvant committed:
removed explicit init of lstm, fstrings instead of format

maxkvant committed:
added normalization,
 moved from SGD to Adam,
 returned explicit init to lstm

maxkvant commented:
> * Some minor violations of pep8

I didn't even know about pep8. I will read it. 

> * Do not upload your checkpoints in repo, since it will become huge. Also take into account that you will have to carry this files in history even after removal. As an alternative you can wget from dropbox or use 
git-lfs.

What do you consider as checkpoints? Binary files for models?

> On the notebook:
> 
> * You load `autoreload` plugin but don't use it

removed that line

> * Please do normalise you features !

added normalization

> * Adam is more safe optimisator choice if you are not planning to spend a lot of time on fine-tuning and lr-scheduling

Adam converges faster, but it overfits faster too. Now, I've switched to Adam, and I wasn't able to reproduce previous accuracy. 

> * For general metrics you might want to use `sklearn.metric`

Yes. I use it except for f1. 
I didn't figure out how to compute f1 for laughter class with function from `sklearn.metric`.

> * Why would you need `global` in jupyter?

Surprisingly, It crashes without `global`.

maxkvant committed:
fix

maxkvant commented:
Did I re-request review on github?

Rauf-Kurbanov reviewed (approved):
Approved, thanks 4/4
